### ISSUE TITLE
Improve errors

### ISSUE THREAD

**Issue by sondrelg** [snok/drf-openapi-tester#213]
## Background 

We currently generate errors with the follow format:

```
E  openapi_tester.exceptions.DocumentationError: Mismatched types, expected a str type.
E  
E  Expected: str
E  
E  Received: {"color": "Red", ...}
E  
E  Sequence: init.array.item.object:key:...
```

This works well in some cases, and seems like a useful format (though we could rework this, if anyone makes a better suggestion) since it provides users with information on:

1. What is wrong
2. What we expected the response data to look like, based on the schema
3. What the response data actually looked like
4. Where in the schema the error is

### Message only

In some cases, like when raising errors related to:
- Any reference handling
- multipleOf
- pattern compiling

The original format breaks down, for many reasons, but the main thing is that points 2-4 from the list above no longer apply. All we need is the message:

```
E  openapi_tester.exceptions.DocumentationError: Expected data to match one or more of the documented anyOf schema types, but found no matches.
```

### Custom expected examples

In other cases, like for:
- pattern validation
- enum validation

Only the second point breaks down. In these cases it would be nice to, not only remove the `expected` attribute, but provide a custom value to represent what we expected based on the schema in these cases. Something like this:

```
E  openapi_tester.exceptions.DocumentationError: The response value is not valid with respect to the enum specified in the schema.
E  
E  Expected: "Cat" or "Dog"
E  
E  Received: "Turtle"
```

## Things to consider

Error messages don't have to end up looking like the examples above, but we should rework the way we raise errors, and we should build in some flexibility to provide different formats, without needing to create new exception types.

We *will* add the possibility of adding custom validators soon, and the custom validators will need to be given the same flexibility our internal validators get. This implicitly makes our error handling a part of the packages public API and means breaking changes to error formatting will require major version bumps.

**Comment by Goldziher** [comment]
I would like to suggest that we remove the error message:

BEFORE
```
E  openapi_tester.exceptions.DocumentationError: Mismatched types, expected a str type.
E 
E  Expected: str
E  
E  Received: {"color": "Red", ...}
E  
E  Sequence: init.array.item.object:key:...
```

AFTER
```
E  openapi_tester.exceptions.DocumentationError: 
E  
E  Expected: str
E  
E  Received: {"color": "Red", ...}
E  
E  Sequence: init.array.item.object:key:...
```

I would also like to suggest we update Sequence - we need to track the property keys more precisely and to be clear this is a map, so instead of init something like

```
Sequence: path:method:array.item:key:... 
```
OR
```
Sequence: path:method:object:key:... 

```

**Comment by sondrelg** [comment]
Yes let's definitely rework the sequence in a separate PR - it's a feature I put in place in 5 minutes that's never really been polished, so I agree there's definitely a lot of room for improvement.

As for removing the messages, I don't have any reservations against it as long as it doesn't make it harder for users to figure out what the problem is. It seems like we're losing information if we *just* remove the messages everywhere. Perhaps we could combine it with more specific error types or something.

**Comment by Goldziher** [comment]
I don't think we need to lose info. Rather, expected and received should be clear - if they are not already. 

I would rename ''sequence'' into something more informative, e.g. ''location'', or ''identifier'' or ''path''.

**Comment by sondrelg** [comment]
Resolved in #216. Reference/sequence will be redone in a separate PR.
